lib/deep_preloader.rb

```diff
@@ -3,129 +3,277 @@
     end
 
-    unloaded_models.each do |m|
-      target = children_by_key[m.read_attribute(parent_key)]
-      association = m.association(association_reflection.name)
-      association.target = target
-      association.set_inverse_instance(target) if target
+    def key
+      model.read_attribute(parent_key_column)
     end
-  end
 
-  def self.preload_multiple_association(models, association_reflection, child_scope, child_preload_spec)
-    assoc_name = association_reflection.name
+    # Conceal the difference between singular and collection associations so
+    # that `load_children` can always `group_by` the key
+    def children
+      target = model.association(association_name).target
 
-    unless association_reflection.macro == :has_many
-      raise "Unsupported association type #{reflection_type}"
+      if collection?
+        target
+      elsif target
+        [target]
+      else
+        []
+      end
     end
 
-    parent_key = association_reflection.active_record_primary_key
-    child_key  = association_reflection.foreign_key
-
-    # some models may have the association already loaded
-    loaded_models, unloaded_models = models.partition { |m| m.association(assoc_name).loaded? }
-    loaded_children = loaded_models.flat_map { |m| m.association(assoc_name).target }
+    def children=(targets)
+      if collection?
+        target = targets
+      else
+        raise "too many children" if targets.size > 1
+        target = targets.first
+      end
 
-    # Load children necessary to resolve unloaded models
-    unloaded_keys = unloaded_models.map { |m| m.read_attribute(parent_key) }
+      ActiveRecord::Base.logger.debug("attaching children to #{model.inspect}.#{association_name}: #{targets}") if DEBUG
 
-    unloaded_children = child_scope.where(child_key => unloaded_keys).to_a
-    unloaded_children_by_key = unloaded_children.group_by { |c| c.read_attribute(child_key) }
+      association = model.association(association_name)
+      association.loaded!
+      association.target = target
+      targets.each { |t| association.set_inverse_instance(t) }
+      targets
+    end
 
-    preload_children = loaded_children + unloaded_children
-    if child_preload_spec.present? && preload_children.present?
-      child_preload_spec.preload(preload_children)
+    def parent_key_column
+      case association_macro
+      when :belongs_to
+        @association_reflection.foreign_key
+      when :has_one, :has_many
+        @association_reflection.active_record_primary_key
+      else
+        raise "Unsupported association type #{@association_reflection.macro}"
+      end
     end
 
-    unloaded_models.each do |m|
-      targets = unloaded_children_by_key.fetch(m.read_attribute(parent_key), [])
-      association = m.association(association_reflection.name)
-      association.loaded!
-      association.target = targets
-      targets.each { |target| association.set_inverse_instance(target) }
+    def child_key_column
```

Comment:
what would it look like? `def dispatch_on_association_direction(when_pointing_to: xxx, when_pointed_from: yyy)`?